packages/core/src/domain/context/trackCustomerData.ts

```diff
@@ -0,0 +1,93 @@
+
+// We observed that the compression ratio is around 8 in general, but we also want to keep a margin
+// because some data might not be compressed (ex: last view update on page exit). We chose 16KiB
+// because it is also the limit of the 'batchBytesCount' that we use for RUM and Logs data, but this
+// is a bit arbitrary.
+export const CUSTOMER_COMPRESSED_DATA_BYTES_LIMIT = 16 * ONE_KIBI_BYTE
+
+export const BYTES_COMPUTATION_THROTTLING_DELAY = 200
+
+export type CustomerDataTracker = ReturnType<typeof createCustomerDataTracker>
+export const enum CustomerDataCompressionStatus {
+  Unknown,
+  Enabled,
+  Disabled,
+}
+
+export function createCustomerDataTracker(
+  type: CustomerDataType,
+  compressionStatus = CustomerDataCompressionStatus.Disabled
+) {
+  let bytesCountCache = 0
+  let alreadyWarned = false
+
+  // Throttle the bytes computation to minimize the impact on performance.
+  // Especially useful if the user call context APIs synchronously multiple times in a row
+  const { throttled: computeBytesCountThrottled, cancel: cancelComputeBytesCount } = throttle((context: Context) => {
+    bytesCountCache = computeBytesCount(jsonStringify(context)!)
+    maybeWarn()
+  }, BYTES_COMPUTATION_THROTTLING_DELAY)
+
+  function maybeWarn() {
+    if (alreadyWarned || compressionStatus === CustomerDataCompressionStatus.Unknown) {
+      return
+    }
+
+    const bytesCountLimit =
+      compressionStatus === CustomerDataCompressionStatus.Disabled
+        ? CUSTOMER_DATA_BYTES_LIMIT
+        : CUSTOMER_COMPRESSED_DATA_BYTES_LIMIT
+
+    if (bytesCountCache > bytesCountLimit) {
+      displayCustomerDataLimitReachedWarning(type, bytesCountLimit)
+      alreadyWarned = true
+    }
+  }
+
+  return {
+    type,
+    updateCustomerData: computeBytesCountThrottled,
+    resetCustomerData: () => {
+      bytesCountCache = 0
+    },
+    getBytesCount: () => bytesCountCache,
+    getCompressionStatus() {
+      return compressionStatus
+    },
+    setCompressionStatus: (newCompressionStatus: CustomerDataCompressionStatus) => {
+      if (compressionStatus === CustomerDataCompressionStatus.Unknown) {
+        compressionStatus = newCompressionStatus
+        maybeWarn()
+      }
+    },
+    stop: () => {
+      cancelComputeBytesCount()
+    },
+  }
+}
+
+const CustomerDataLabel = {
+  [CustomerDataType.FeatureFlag]: 'feature flag evaluation',
+  [CustomerDataType.User]: 'user',
+  [CustomerDataType.GlobalContext]: 'global context',
+  [CustomerDataType.LoggerContext]: 'logger context',
+}
+
+function displayCustomerDataLimitReachedWarning(customerDataType: CustomerDataType, bytesCountLimit: number) {
+  display.warn(
+    `The ${CustomerDataLabel[customerDataType]} data exceeds the recommended ${
+      bytesCountLimit / ONE_KIBI_BYTE
+    }KiB threshold. More details: https://docs.datadoghq.com/real_user_monitoring/browser/troubleshooting/#customer-data-exceeds-the-recommended-3kib-warning`
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I don't know it you already have it in mind but we should probably revisit the documentation to take the compression limit into account and also advice customer to enable the compression if they haven't already.

Reply:
Yes, I was planning to update the docs too.

Reply:
Nice, so the link might also change as it includes the limit "#customer-data-exceeds-the-recommended-**3kib**-warning"